lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php

```diff
@@ -217,18 +235,37 @@ protected function connectTo($connectionName)
     }
 
     /**
-     * @param string $connectionName
-     * @param array  $params
+     * Connets to a random slave and tries others until if found a server that is alive
      *
-     * @return mixed
+     * @param array $params
+     * @param array $driverOptions
+     *
+     * @return Driver\Connection
+     *
+     * @throws ConnectionException
      */
-    protected function chooseConnectionConfiguration($connectionName, $params)
+    private function connectToSlave($params, $driverOptions)
     {
-        if ($connectionName === 'master') {
-            return $params['master'];
+        shuffle($params['slaves']);
```

Comment:
This is a replacement of the `array_rand()` stuff that was done before. See removed diffs